### ISSUE TITLE
resource.Test cleans up module files before the test step has a chance to load templates from disk

### ISSUE THREAD

**Issue by sjauld** [hashicorp/terraform#17304]
We are currently using the resource.Test function to run acceptance tests over whole terraform modules. This works great normally, however when our module contains a template, the template is not available on disk during the test steps and everything fails.

### Terraform Version
```
v0.11.3
```

### Terraform Configuration Files
```hcl
# module/main.tf
variable "test" {}

resource template_file "test" {
  template = "${file("${path.module}/test.yml.tpl")}"

  vars {
    my_var = "${var.test}"
  }
}
```

```yml
# module/test.yml.tpl
var: ${my_var}
```

```go
# tests/test_init.go
package tests

import (
	"log"
	"os"
	"testing"

	"github.com/hashicorp/terraform/helper/schema"
	"github.com/hashicorp/terraform/terraform"
	"github.com/terraform-providers/terraform-provider-template/template"
)

var testAccProviders map[string]terraform.ResourceProvider
var testAccTemplateProvider *schema.Provider

func init() {
	testAccTemplateProvider = template.Provider().(*schema.Provider)
	testAccProviders = map[string]terraform.ResourceProvider{
		"template": testAccTemplateProvider,
	}
}
```

```go
# tests/module_test.go
package tests

import (
	"fmt"
	"testing"

	"github.com/hashicorp/terraform/helper/acctest"
	"github.com/hashicorp/terraform/helper/resource"
)

const configSourceDir = "/config"

func TestAccModule(t *testing.T) {
	resource.Test(t, resource.TestCase{
		Providers: testAccProviders,
		Steps: []resource.TestStep{
			{
				Config: testAccModule(),
			},
		},
	})
}

func testAccModule() string {
	return fmt.Sprintf(`
module "test" {
  source = "%s/module"

  test = "potato"
}
`, configSourceDir)
}

```

### Debug Output
N/A

### Crash Output
N/A

### Expected Behavior
`TF_ACC=1 go test -v -run TestAccModule ./... ` should run as normal and not fail

### Actual Behavior
```
testing.go:513: Step 0 error: config is invalid: 1 problem:

		- module.test.template_file.test: 1 error occurred:

		* module.test.template_file.test: file: open /tmp/tf-test090410664/.tfmodules/4fb9789cb9421ced3f4a4d36cc618680/test.yml.tpl: no such file or directory in:

		${file("${path.module}/test.yml.tpl")}
```

### Steps to Reproduce
It's complicated - we bundle our tests into a Docker container (https://hub.docker.com/r/cozero/tf-acceptance-testing/) containing all the source code required, then run something like 
```
docker run --rm \
    -e TF_ACC=1 \
    -v $(pwd):/config \
    -v $(pwd)/tests:/go/src/app/tests \
    cozero/tf-acceptance-testing:v0.11.3 test -v -run ./...
```
to run the acceptance tests.

### Additional Context

helper/resource/testing.go:745 basically wipes our config files off the disk before the test steps are executed, meaning that any template files are unavailble, causing the failure.

If we comment this line out, we can successfully run our tests, however obviously this leaves unwanted tmp files behind on the disk.

Not sure if maybe removing these files can be moved to a global cleanup step? 

### References

N/A

**Comment by apparentlymart** [comment]
Hi @sjauld! Sorry for this odd behavior.

The test utilities in `helper/resource` are mainly there to support the Terraform provider acceptance tests, and so they probably have a number of limitations that make them harder to use for other purposes like what you are doing here. In particular, acceptance tests do not use child modules, and so the system isn't really designed with that in mind.

With that said, your testing use-case is something we've had on our list to support eventually via CLI commands (as opposed to using the `resource.TestStep` API directly), and if we can make something work for you now without a lot of reorganization then we're open to it.

Unfortunately the Terraform team at HashiCorp currently has its attention focused elsewhere (on the configuration language) so we won't be able to look deeper into this at the moment, but if you have some specific ideas about what could be done here I'd be open to discuss them here so we can gather the feedback to inform some later work here.

**Comment by sjauld** [comment]
Hi @apparentlymart,

Appreciate that you're hard at work on HCL2 - we're looking forward to using it!

I have just added a PR with a proposed workaround (#17327) - it's not ideal since it will leave the temp files around on disk, but it works for our use case since we're running in docker and nuking the container every time anyway. 

Let if me know if this is on the right track and I can make any improvements required to get this safely into master!

Cheers,

**Comment by sjauld** [comment]
Closed by #17327
